fluidRow(
  mainPanel(
    h1("Kappaphycus alvarezii gene-set analysis"),
    p("This is an R Shiny App to facilitate gene set analysis and annotation exploration of Kappaphycus alvarezii sequencing data"),
    
    h2("RNA-seq Pre-processing"),
    p("For pre-procesiing of your own RNA-seq data, please visit our ", a("Galaxy", href="https://galaxy.forjazul.com/"), " platform that offers automated pipelines to estimate transcript counts"),
    
    h2("Data import"),
    p("This app takes as input transcript counts. This can be either generated by the submission of RNA-Seq data to our adapted Galaxy pipeline ot be using pre-compiled data available by us."),
    
    h3("User's data (Galaxy)"),
    p("This App is connected direclty with the user's Galaxy output repository. Thus, if a user wants to analyse his private data, that will be available under the Custom import option and by providing the username"),
    
    h3("MSI data"),
    p("MSI data has been already pre-compiled by us and is available under the Pre-compiled import option"),
    
    h3("Publicly available data from NCBI"),
    p("NCBI available sequence data has been already retrieved and pre-compiled by us and is available under the Pre-compiled import option"),
    
    h2("Metadata import"),
    p("some text here"),
    
    h2("Differential expression analysis"),
    p("Upon user's selection of input data and configuration of the metadata information, the app will prompt the user to select the available treatmens or subgroups to be used for differential expression analysis. 
      The users have the option to select more than one contrast"),
  
    
    
    h2("Overrepresentation analysis (ORA)"),
    p("Over-representation analysis (ORA) is a statistical method used to identify significantly enriched biological pathways or gene sets within a given RNAseq dataset based on user defined fold-change and FDR. 
    This method looks at the frequencies of genes within a particular pathway or gene set and compares them to the frequencies of genes in the overall dataset to determine if the pathway or gene set is significantly enriched."),
    
    h2("Gene Set Enrichment Analysis (GSEA)"),
    p("GSEA, or gene set enrichment analysis, is another statistical method used to identify significantly enriched pathways or gene sets within a given dataset. Unlike RNAseq over-representation analysis, 
      GSEA takes into account the relative expression levels of genes within a pathway or gene set and compares them to the expression levels of genes in the overall dataset. 
      This allows GSEA to identify pathways or gene sets that are not only significantly enriched, but also have a consistent trend of either upregulation or downregulation."),
    
    h2("Clustering analysis"),
    p("Clustering algorithms group genes with similar expression patterns together, allowing researchers to identify clusters of genes that may be involved in the same biological pathway or process. User defined fold-change and FDR."),
    
    h2("KEGG pathways analysis"),
    p("Kegg pathway analysis is a useful tool for identifying significant pathways and functional pathways in RNAseq data. 
      The analysis is typically performed by comparing the fold-change in gene expression between two different experimental conditions. 
      This allows researchers to identify which pathways are significantly differentially expressed between the two conditions, 
      and to further investigate the functional significance of these pathways.")
  )
  
)
